### PR TITLE
VB-2445 - Validations around session time and valid from updates.

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Adds validations around updates to session times and valid from date.
## What is the intent behind these changes?

Admin API improvements.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt [hunk 1]
```diff
@@ -93,35 +96,23 @@ interface SessionTemplateRepository : JpaRepository<SessionTemplate, Long> {
   @Modifying
   fun deleteByReference(reference: String): Int
 
-  @Modifying
+  @Modifying(clearAutomatically = true)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:99]: have not used clearAutomatically = true before

**Review by AledWEvans — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:99]: tests were failing as the entity was not being refreshed after updating name.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt [hunk 2]
```diff
@@ -165,25 +165,45 @@ class SessionTemplateService(
...
+  private fun validateUpdateSessionTemplate(reference: String, updateSessionTemplateDto: UpdateSessionTemplateDto) {
+    val sessionTemplate = SessionTemplateDto(getSessionTemplate(reference))
+    val hasVisits = visitRepository.hasVisitsForSessionTemplate(reference)
+    validateUpdateSessionTemplateTime(sessionTemplate, updateSessionTemplateDto, hasVisits)
+    validateUpdateSessionTemplateFromDate(sessionTemplate, updateSessionTemplateDto, hasVisits)
+  }
+
+  private fun validateUpdateSessionTemplateTime(existingSessionTemplate: SessionTemplateDto, updateSessionTemplateDto: UpdateSessionTemplateDto, hasVisits: Boolean) {
+    // if a session has visits its time cannot be updated.
+    if (updateSessionTemplateDto.sessionTimeSlot != null && existingSessionTemplate.sessionTimeSlot != updateSessionTemplateDto.sessionTimeSlot && hasVisits) {
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/SessionTemplateService.kt:177]: Yes.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/SessionDateRangeDto.kt [hunk 3]
```diff
@@ -5,7 +5,7 @@ import jakarta.validation.constraints.NotNull
 import uk.gov.justice.digital.hmpps.visitscheduler.controller.validators.SessionDateRangeValidation
 import java.time.LocalDate
 
-class SessionDateRangeDto(
+data class SessionDateRangeDto(
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/SessionDateRangeDto.kt:8]: Yes I should have created it as a data class from the start.

**Pull request merged**